### PR TITLE
Added CUDAHOSTCXX variable needed to compile with cuda and mpi.

### PR DESCRIPTION
**Pull request by frobnitzem**
Spack seems to set the CUDAHOSTCXX environment variable in the build environment, which cmake picks up and passes to nvcc as -ccbin=(...)  Mine was g++-6.

When compiling with mpi, cuda files cannot include <mpi.h> unless the host compiler points to the mpi wrapper.  Without this environment variable specifically set, no amount of cmake options could switch over the -ccbin from g++-6.

Shouldn't spack automatically set CUDAHOSTCXX to the right thing when using mpi?  Maybe I missed some documentation on this.

### CODE REVIEW

**Comment by adamjstewart** [comment]
Maintainers: @davydden @jppelteret @luca-heltai 
Cuda experts: @svenevs @ax3l

**Comment by svenevs** [comment]
I know very little about mpi, does it work if you remove CUDAHOSTCXX entirely (here and in CUDA or just to test I think you can unset it)?

Ref: https://github.com/spack/spack/pull/16869#issuecomment-642318695

If removing it works, maybe we should just not set it? Or somehow not set it for mpi? Is this a wrapper and if so can we rely on it wrapping the right stuff as long as they show up in the PATH?

Unsetting may interfere with the LLVM as CUDA plans though. Unfortunately I don't have much else to offer other than potential workarounds and the original discussion.  The current plan of overwriting it seems fishy to me, but I really have no idea :confused:

**Comment by jppelteret** [comment]
Ping @Rombur, who is our CUDA expert. I must admit that I have experience with CUDA, so I'm not qualified to say whether or not this change is necessary or the correct one to make. But if it is, then the modification looks fine to me!

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 1]
```diff
@@ -529,3 +529,8 @@ def cmake_args(self):
 
     def setup_run_environment(self, env):
         env.set('DEAL_II_DIR', self.prefix)
+
+    def setup_environment(self, spack_env, run_env):
+        spec = self.spec
+        if '+mpi' in spec:
```
Comment (jppelteret) [inline var/spack/repos/builtin/packages/dealii/package.py]: I suppose that this should only be done if `+cuda` is in spec as well, i.e.
`if '+cuda' in spec and '+mpi' in spec:`

**Comment by Rombur** [comment]
Sorry I've never used `CUDAHOSTCXX` before. This change looks fine to me but, like I said, I don't have any experience using this variable.

**Comment by frobnitzem** [comment]
I see two issues here - first is a big picture problem with how to model hierarchies of compilers.

The way I usually use mpicxx is to set OMPI_CXX=g++-6 (or other underlying compiler), so that mpicxx adds its flags and calls the right thing. At present, the openmpi package does this in its setup_dependent_build_environment.  I think this is a good basic working policy to setup hierarchies of compilers and is worth documenting.

However, it doesn't handle advanced cases where a package like dealii wants to use swap cuda's underlying compiler with mpic++ after cuda has already been configured.  One could also imagine the reverse situation where a package wants mpic++ on top of nvcc.  I don't know how spack would support that (maybe ^mpi backend=cuda ^cuda@10 backend=gcc@6 or %mpi:cuda:gcc).

The other (simpler) issue with this build is that I tried adding "-DCUDA_HOST_COMPILER=mpic++" to cmake flags and it had no effect because the environment variable took precedence.  This could be a problem with the package's CMakeLists.txt scripts, or with stock cmake-3.17 I'm using since the [cmake docs](https://cmake.org/cmake/help/v3.17/module/FindCUDA.html) are ambiguous about whether CUDA_HOST_COMPILER flag should take precedence over the env variable, [CUDAHOSTCXX](https://cmake.org/cmake/help/v3.17/envvar/CUDAHOSTCXX.html).

Specifically, using flags `-DCMAKE_CUDA_HOST_COMPILER=mpic++ -DCUDA_HOST_COMPILER=mpic++`, CMakeCache.txt contains both:
```
CMAKE_CUDA_HOST_COMPILER:UNINITIALIZED=/usr/bin/mpic++
CUDA_HOST_COMPILER:FILEPATH=/usr/bin/mpic++
```
but the nvcc invocations are nevertheless all `-ccbin=/usr/bin/g++-6`.

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 1 above)
Comment (frobnitzem) [reply var/spack/repos/builtin/packages/dealii/package.py]: I updated this to unset the CUDAHOSTCXX variable, which seems like a less invasive change for this package (and still works).

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 2]
```diff
@@ -387,6 +387,11 @@ def cmake_args(self):
                 '-DMPI_CXX_COMPILER=%s' % spec['mpi'].mpicxx,
                 '-DMPI_Fortran_COMPILER=%s' % spec['mpi'].mpifc,
             ])
+            if '+cuda' in spec:
+                options.extend([
+                    '-DDEAL_II_MPI_WITH_CUDA_SUPPORT:BOOL=ON',
+                    '-DCUDA_HOST_COMPILER=%s' % spec['mpi'].mpicxx
+                ])
```
Comment (jppelteret) [inline var/spack/repos/builtin/packages/dealii/package.py:401]: I wasn't aware of the [`DEAL_II_MPI_WITH_CUDA_SUPPORT`](https://github.com/dealii/dealii/blob/159efe444170a59298b43472bbf7184028b735ab/cmake/configure/configure_1_mpi.cmake#L60) flag. Glad that you found it!

In #19253, we're updating the package to use a newer syntax that the `CMakePackage` offers. Perhaps you'd be willing to follow suite? So I think that this block would then read
```
options.extend([
    self.define('DEAL_II_MPI_WITH_CUDA_SUPPORT, True),
    self.define('CUDA_HOST_COMPILER', spec['mpi'].mpicxx)
])
```

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 3]
```diff
@@ -529,3 +534,9 @@ def cmake_args(self):
 
     def setup_run_environment(self, env):
         env.set('DEAL_II_DIR', self.prefix)
+
+    def setup_build_environment(self, env):
+        spec = self.spec
+        if '+cuda' in spec and '+mpi' in spec:
+            env.unset('CUDAHOSTCXX')
+        #    #env.set('CUDAHOSTCXX', spec['mpi'].mpicxx)
```
Comment (jppelteret) [inline var/spack/repos/builtin/packages/dealii/package.py]: I suppose this line can be removed now.

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 4]
```diff
@@ -387,6 +387,11 @@ def cmake_args(self):
                 '-DMPI_CXX_COMPILER=%s' % spec['mpi'].mpicxx,
                 '-DMPI_Fortran_COMPILER=%s' % spec['mpi'].mpifc,
             ])
+            if '+cuda' in spec:
+                options.extend([
+                    '-DDEAL_II_MPI_WITH_CUDA_SUPPORT:BOOL=ON',
+                    '-DCUDA_HOST_COMPILER=%s' % spec['mpi'].mpicxx
```
Comment (jppelteret) [inline var/spack/repos/builtin/packages/dealii/package.py]: I could be mistaken, but I think that this should be `CMAKE_CUDA_HOST_COMPILER`? See https://gitlab.kitware.com/cmake/cmake/-/issues/19013

Comment (jppelteret) [reply var/spack/repos/builtin/packages/dealii/package.py]: Ah, I now see in the documentation that you [mentioned](https://cmake.org/cmake/help/v3.17/module/FindCUDA.html) that `CUDA_HOST_COMPILER` is listed. I wonder which one is indeed correct?

**Comment by jppelteret** [comment]
> Sorry I've never used CUDAHOSTCXX before. This change looks fine to me but, like I said, I don't have any experience using this variable.

Thanks a lot for your input @Rombur!

In the documentation that @frobnitzem [mentioned](https://cmake.org/cmake/help/v3.17/module/FindCUDA.html) they state that the `FindCUDA` module has been deprecated since v3.10. But I see that we unconditionally call `FindPackage(CUDA)` [here](https://github.com/dealii/dealii/blob/a975b5dc58cd002020110294d70be0f07a9ca2ed/cmake/configure/configure_1_cuda.cmake#L38). Which version of CMake are you using? *Maybe* that's got something to do with the problem that you're seeing here -- I'm 100% speculating, but it could be that the setup in the later `FindCUDAToolkit` is slightly different and clashes with something that `FindCUDA` is doing. Either way, if the workaround fixes the issue for you then I'd be happy for us to take it for now. If it has any unintended side-effects for other users then we might have to revert the change and figure something else out.

**Comment by adamjstewart** [comment]
> Ping @Rombur, who is our CUDA expert.

@Rombur would you be interested in helping to maintain Spack's CUDA package and base class? We could really use more people in charge of reviewing PRs/tackling issues.

**Comment by svenevs** [comment]
TL;DR see point 5 enumerated below.

https://github.com/dealii/dealii/blob/d949861f2fb94acc0ce0f318983e214f87913c34/CMakeLists.txt#L36

You're stuck in the black hole that is cmake/cuda.  Until you adopt [at least CMake 3.8](https://developer.nvidia.com/blog/building-cuda-applications-cmake/) (which is going to be a HUGE task for you all), **AND** use the `CUDA` language support, you won't be able to take advantage of many of the newer features.  Any of the `CMAKE_CUDA_<anything>` variables need you to enable the CUDA language, the leading `CMAKE_` being the key.  For your current setup, you need to be forcibly setting the old [`FindCUDA`](https://cmake.org/cmake/help/latest/module/FindCUDA.html) variables, e.g., `CUDA_HOST_COMPILER` (no leading `CMAKE_`).  Cliff notes (since this stuff is super confusing):

1. You're using deprecated CMake behavior, the technical debt incurred by not updating will continue to grow.  Advise switching to 3.18 and if you hit problems this is an important project that Kitware and the devs will likely be very interested in getting working.  The below information is included hopefully for your benefit in assessing as a project what you will want to do.  cpplang slack #cmake and #cuda channels are good places to get help if stuck, and if issues arise report them to the Kitware tracker.
2. For the purposes of this PR, any testing should be manipulating the old `FindCUDA` variables (**no** `CMAKE_` prefix).
3. In the context of spack, all of the intricacies of `FindCUDA` vs `CUDA` language are not accounted for, because it's rather complicated.  We are setting `CUDAHOSTCXX` in the first place because most projects are still using `FindCUDA`, and it will consider this variable specifically.  In the new world order with the CUDA language enabled, you can set `CMAKE_CUDA_COMPILER` and `CMAKE_CUDA_HOST_COMPILER` respectively.
4. If you update your CMake, don't forget that you will now need to conditionally set different variables in this spack package (sigh).
5. So AFAICT upon closer inspection of your build system code, while it is "fishy" in the CMake sense, this appears to be the right thing to do:

    ```py
    def setup_build_environment(self, env):
        spec = self.spec
        if '+cuda' in spec and '+mpi' in spec:
            env.set('CUDAHOSTCXX', spec['mpi'].mpicxx)
    ```

---------------------------

Some suggestions on revamping your CMake // CUDA setup:

```cmake
if (CMAKE_VERSION VERSION_LESS 3.18.0 AND ${some_option_enabling_cuda})
  message(FATAL_ERROR "To use dealii and CUDA you must have CMake 3.18+")
endif()
if (${some_option_enabling_cuda})
  cmake_policy(3.18.0)  # VERY important hack, overrides cmake_minimum_required which sets policy stack
  include(CheckLanguage)
  check_language(CUDA)
  if (NOT CMAKE_CUDA_COMPILER)
    message(FATAL_ERROR "Could not find suitable CUDA compiler")
  endif()
endif()
```

I do not remember if you can conditionally set `cmake_minimum_required`, but I know you cannot use the `project(dealii VERSION x.y.z LANGUAGES ... CUDA ...)` conditionally (aka `check_language` required if CUDA support optional) because CMake requires `project(...)` calls to be "unguarded" (not in an `if`, not macro, etc).  If you're going to spend any time revisiting your cmake / cuda stuff, it is in your interest to adopt 3.18 (not 3.8).  3.18 specifically introduces clang as cuda compiler, but also a host of other goodies like [`CMAKE_CUDA_ARCHITECTURES`](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/4568) and it seems like also some MPI fixes which are likely relevant to you.  That said, it's not clear to me where separable and/or ptx compilation are and so it may not be suitable for you if you're needing to change the CUDA compiler from `nvcc` to something not officially supported.  It looks like you can [set `CMAKE_CUDA_COMPILER_FORCED=ON`](https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/CMakeTestCUDACompiler.cmake#L4), but there are things going on with [`FindMPI.cmake` and flag manipulation based of language CUDA and compiler id](https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/FindMPI.cmake#L1157) that may be problematic depending on how you are wrapping things.

It's probably also worth mentioning, if you do this, nobody has any room to complain about 3.18 if you only force it on them for CUDA.  "But I can't install it through `apt` because I'm on Debian".  Use `spack`, or `pip`, or anything else.  A new enough CMake not being available on the system is not an excuse (/rant).

---------------------------

I hope at least some of this is useful, I sent you in the wrong direction and feel bad about that.  Legacy CMake hacks on CUDA compilation are always quite varied (and also very common, because you gotta do what you gotta do).  I should have looked at your build system first.

> would you be interested in helping to maintain Spack's CUDA package and base class? We could really use more people in charge of reviewing PRs/tackling issues.

To clarify, help is needed from somebody who knows CUDA.  Adam and the other contributors are happy to answer "spack" questions related to the build system (how do the packages inherit, etc) on the spack slack channel, but we need help from people who know CUDA.  I am having significant personal problems and need to step down for a while and try and dig myself out, as what little time I can spend on this is (as evidenced by the above) becoming counterproductive.  But we only have one other maintainer, and he really needs additional support.  Thanks to anybody who is willing to step up :heart: :pray:  Whoever does step up, since I know the CMake side, will in exchange have license to ask me any CMake / CUDA questions behind the scenes if that sweetens the deal enough?

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 4 above)
Comment (frobnitzem) [reply var/spack/repos/builtin/packages/dealii/package.py]: I used cmake 3.17, and CUDA_HOST_COMPILER was picked up by more parts of cmake, but still didn't completely work - see discussion below.

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 2 above)
Comment (Rombur) [reply var/spack/repos/builtin/packages/dealii/package.py:401]: I don't know if you want to set that flag (`DEAL_II_MPI_WITH_CUDA_SUPPORT`). When that flag is ON we assume that MPI was compiled with CUDA support. This avoids copy to the host but if MPI was compiled without CUDA support the code will crash.

**Comment by Rombur** [comment]
@adamjstewart Sure, I can help.

@svenevs Thanks for the detailed explanation. Our build system is in a strange state and to configure with CUDA we require CMake 3.9. We added CUDA support a long time ago and like you mentioned, a lot of things were changed with CMake 3.8. So, we decided to require CMake 3.9 and to update most of the CUDA build system but we did not update all of it that's why we still use `FindPackage(CUDA)`. If I remember correctly there were some variables that were easier to grab with `FindPackage(CUDA)`, so we have this strange thing where we set CUDA as a language but we still call `FindPackage(CUDA)`. 

It's on my todo list to rework our build system because I really want CUDA clang :smiley:

**Comment by svenevs** [comment]
Thanks so much!  I'm signed into (but inactive) in spack and cpplang slack channels and can definitely catch you up on the brave new world of CMake + CUDA, a lot of people really stepped up to get it there and it's really quite spiffy and nice :slightly_smiling_face:  But since the CMake docs are really more of a reference, it's difficult to know what to even look for :smirk: Most likely, the things you were needing are now available.  If slack does not work for you, send me an e-mail with cmake / cuda or something catchy in the title so I don't miss it.

**Comment by ax3l** [comment]
> Shouldn't spack automatically set CUDAHOSTCXX to the right thing when using mpi? Maybe I missed some documentation on this.

Hm, I am setting this since recently in dependent build environments of `CudaPackage`: #16869 (note regression in #17823/#17826).

Setting this to `spec['mpi'].mpicxx` is a bit unusual... looks like a bug in the CMake scripts to me. E.g. a missing `FindPackage(MPI)` and "CMake-link" of its respective targets (`target_link_libraries(<target-here> PUBLIC MPI::MPI_C MPI::MPI_CXX)`)?

**Review by ax3l — APPROVED** [review]
Besides the hint above, I think your current work-around is legit until CMake scripts are modernized and fabulous :sparkles:

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 2 above)
Comment (frobnitzem) [reply var/spack/repos/builtin/packages/dealii/package.py:401]: This project moves really fast - making git rebase indispensable.  This update uses the new syntax, and I've tested it's working on my system.

I should note that I monkey-patched Sacado locally to get this to compile (https://github.com/trilinos/Trilinos/issues/8174), but that's not deal.II's fault.

**Comment by jppelteret** [comment]
@Rombur I see that the flag for `DEAL_II_MPI_WITH_CUDA_SUPPORT` is now set to `true`. IIRC this wasn't enabled in the first version of the patch opened in the PR. Is this going to be a problem in general, i.e. would most users using the `+mpi +cuda` variant to keep it disabled? Ideally we'd want the check to be built in to our cmake scripts, but I guess that's on your/Daniel's/Matthias' todo list.

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 2 above)
Comment (jppelteret) [reply var/spack/repos/builtin/packages/dealii/package.py:401]: > This project moves really fast - making git rebase indispensable. This update uses the new syntax, and I've tested it's working on my system.

You were just a bit unlucky that I happened to pick this weekend to make a few touch-ups, and then that snowballed into a complete revamp of the package :-D Thanks for updating the changes that you made, and for retesting them!

**Comment by Rombur** [comment]
>  Is this going to be a problem in general, i.e. would most users using the +mpi +cuda variant to keep it disabled? 

If you have compiled MPI with CUDA support, you want the flag otherwise you cannot use it. I think it would be safer to revert that change.

> Ideally we'd want the check to be built in to our cmake scripts, but I guess that's on your/Daniel's/Matthias' todo list. 

Unfortunately, it's not and I haven't found any project that deal with this nicely. The problem is that CUDA support is not part of the MPI standard and so every MPI implementation deals with CUDA support in a different way. Last time I checked, MPICH did not support CUDA, OpenMPI had to be compiled with CUDA support, and Spectrum MPI was using a runtime flag...

**Comment by jppelteret** [comment]
>  If you have compiled MPI with CUDA support, you want the flag otherwise you cannot use it. I think it would be safer to revert that change.

Thanks, that's what I thought. @frobnitzem, would you have any objection to me doing this? Do you specifically require this option to be set? If so then it might be that, for now, you'd have to patch this locally OR we could also add another variant that specifically sets the flag.

> Unfortunately, it's not and I haven't found any project that deal with this nicely. The problem is that CUDA support is not part of the MPI standard and so every MPI implementation deals with CUDA support in a different way. Last time I checked, MPICH did not support CUDA, OpenMPI had to be compiled with CUDA support, and Spectrum MPI was using a runtime flag...

That's a shame :-/

**Comment by frobnitzem** [comment]
Not reading the docs on DEAL_II_MPI_WITH_CUDA_SUPPORT is my fault.  I've created #19279 to make that flag more conditional.
